compiler/rustc_middle/src/mir/query.rs

```diff
@@ -233,8 +233,9 @@ pub enum ConstraintCategory<'tcx> {
     UseAsConst,
     UseAsStatic,
     TypeAnnotation,
-    Cast {
-        /// Whether this is an unsizing cast and if yes, this contains the target type.
+    Cast,
```

Comment:
Why does a cast not have `unsize_to`? Casts can be unsizing, for example: `&() as &dyn std::any::Any`.

Reply:
The intent was to have `Cast` represent only the non-coercion casts from the table [here](https://doc.rust-lang.org/nightly/reference/expressions/operator-expr.html#type-cast-expressions)[^1] and use `Coercion` for any coercion whether they are caused by a cast or not. But seeing that MIR also just represents coercions as special kind of cast and not the other way around, maybe we should just add a field `is_coercion` to the `Cast` variant and not add a new variant.

[^1]: side note: function item to function pointer and closure to function pointer should not be listed there as they are coercions.

Reply:
> maybe we should just add a field `is_coercion` to the `Cast` variant and not add a new variant.

This seems appropriate. I think it's nice to mention what the user actually written in syntax. But trying to explain that `as` casts are a super set of coercions is probably confusing.